pkg/kubelet/network/cni/cni.go

```diff
@@ -43,6 +44,8 @@ type cniNetworkPlugin struct {
 	host           network.Host
 	execer         utilexec.Interface
 	nsenterPath    string
+	mu             sync.Mutex
+	podMutexes     map[string]*sync.Mutex
```

Comment:
The map-of-mutexes is because there's nothing in the CNI plugin itself that demands that we block pod A network setup because pod B is being set up.  But we clearly should block Pod A teardown on Pod A setup.

Basically, same-pod operations should probably be locked.  But there's no reason to lock operations between separate pods.  If the specific CNI plugin needs that, it can handle that itself internally.  But kube shouldn't do it.

Yes, this could be pushed up the stack, but that would require some significant re-work of runtimes.  At least the docker runtime is currently quite happy to start a teardown while a setup is ongoing, last I looked a couple months ago.  The main problem is kubelet.go::killPod(), which is called from the pod sync goroutine.  It's also called from podKiller() on a separate goroutine.  If you create a pod and then kill it within 10 seconds or so, I've seen setup/teardown race with each other.